File: LeetCode/Medium/Minimize The Maximum Difference Of Pairs/C++/min_max_diff.cc
class Solution {
public:
    int count_valid(vector<int> nums, int mid) {
        int ind = 0;
        int cnt = 0;
        while (ind < nums.size()-1) {
            if (nums[ind+1]-nums[ind] <= mid) {
                cnt++;
                ind++;
            }
            ind++;
        }
        return cnt;
    }

    int minimizeMax(vector<int>& nums, int p) {
        sort(nums.begin(), nums.end());
        int l=0, h=nums[nums.size()-1]-nums[0];
        while(l < h) {
            int mid=l+(h-l)/2;
            if (count_valid(nums, mid) >= p) h = mid;
            else l=mid+1;
        }
        return l;
    }
};

File: LeetCode/Medium/Minimize The Maximum Difference Of Pairs/Java/minMaxDiff.java
class Solution {
    public int minimizeMax(int[] nums, int p) {
        Arrays.sort(nums);
        int n = nums.length;
        int l=0, h=nums[n-1]-nums[0];
        while(l < h) {
            int mid = l+(h-l)/2;
            if (countValidPairs(nums, mid) >= p) h = mid;
            else l = mid+1;
        }
        return l;
    }

    private int countValidPairs(int[] nums, int thres) {
        int ind=0, cnt=0;
        while(ind < nums.length-1) {
            if (nums[ind+1] - nums[ind] <= thres) {
                cnt++;
                ind++;
            }
            ind++;
        }
        return cnt;
    }
}
